### PR TITLE
trhooks: Add TaskStopHook interface to services

### LINKED ISSUE #5770: Nomad not cleaning consul services/healthcheck after failed deploy [0.9.x]
### Nomad version
Noticed on 0.9.0. Problem present on 0.9.2-rc1 and 0.9.1 as well

```
nomad version
Nomad v0.9.2-rc1 (98e75251316173bd2e6f6392dceea5f2d2182e17)
```

### Operating system and Environment details
Running on Ubuntu
Nomad single node, both client and server
Consul single node, both client and server (v1.4.4) - tested with v1.5.1 as well
Vault single node (v1.1.0)
Tried with and without ACLs enabled (both nomad and consul)

### Issue
When provisioning a job, that contains a template, referencing to a vault secret, the following actions happen:

* Deployments start
* job gets scheduled, task gets allocated
* Service gets registered in consul
* Nomad deployment fails, with error:

```
Template failed: vault.read(aws/creds/my-creds): vault.read(aws/creds/my-creds): Error making API request. URL: GET http://127.0.0.1:8200/v1/aws/creds/my-creds Code: 403. Errors: * 1 error occurred: * permission denied
```
* Healthcheck is still registered in consul, failing

### Reproduction steps
* Provide a cluster with a running nomad, consul and vault
* Run nomad, ensure it's registered properly in consul
* Try `nomad run my-job.nomad` (see file given below).
* You should see nomad registering a first healthcheck as failing in consul (all good here)
* Allocation fails 
* See consul services, that still lists the service with a failing healthchecks.

### Job file (if appropriate)
This has been tested and reproduced with the following job file:

```
job "my-task2" {
  datacenters = ["mickael"]
  type = "service"
  group "my-task2" {
    count = 1
    task "my-task" {
      driver = "docker"
      shutdown_delay = "10s"
      resources {
        memory = 1500
        cpu = 2000
        network {
          port "service_port" {
          }
        }
      }

      template {
        data = <<EOH
          {{ with secret "aws/creds/my-task" }}
AWS_ACCESS_KEY_ID={{ .Data.access_key }}
AWS_SECRET_ACCESS_KEY={{ .Data.secret_key }}
          {{ end }}
        EOH
        destination = "secrets/creds.env"
        env         = true
        change_mode = "restart"
      }

      config {
        image        = "nginx"
        port_map {
          service_port = 80
        }
        labels {
          build_id = "1"
        }
        dns_servers = [
          "172.17.0.1"
        ]
      }

      service {
        name = "my-task"
        tags = [
                 "nomad",
                 "my-task",
          ]
        port = "service_port"
        check {
          name     = "serviceHealth_my-task"
          type     = "http"
          path     = "/"
          interval = "10s"
          timeout  = "2s"
        }
      }

      vault {
        policies = [
                    "nomad_my-task",
                   ]
        change_mode = "restart"
      }
    }
  }
}
```

Also reproduced with a custom upgrade/reschedule policy, which lead to two allocation in total, and to two failing healthchecks registered in consul. (We could match the consul HC with nomad allocations via the port registered.)

```
    reschedule {
        attempts       = 15
        interval       = "30m"
        delay          = "5s"
        delay_function = "exponential"
        max_delay      = "180s"
        unlimited      = false
    }

    update {
      max_parallel = 1
      min_healthy_time = "10s"
      healthy_deadline = "570s"
      progress_deadline = "575s"
      auto_revert = true
    }
```

### Logs
Not much information was found in the logs.
Same for consul, logs are clean.
[nomad.log](https://github.com/hashicorp/nomad/files/3248344/nomad-issue.log)

Consul configured:
* With/without ACLs whitelist mode (tested both)
* with encryption

Nomad configured:
* With/Without ACLs (tested both)
* TLS enabled
* with encryption
* vault and consul integration enabled

### PR DESCRIPTION
**Pull request by endocrimes**
We currently only run cleanup Service Hooks when a task is either
Killed, or Exited. However, due to the implementation of a task runner,
tasks are only Exited if they every correctly started running, which is
not true when you recieve an error early in the task start flow, such as
not being able to pull secrets from Vault.

This updates the service hook to also call consul deregistration
routines during a task `Stop` lifecycle event, to ensure that any
registered checks and services are cleared in such cases.

fixes #5770

### CODE REVIEW

File: client/allocrunner/taskrunner/service_hook.go [hunk 1]
```diff
@@ -179,6 +182,17 @@ func (h *serviceHook) deregister() {
 
 }
 
+func (h *serviceHook) Stop(ctx context.Context, req *interfaces.TaskStopRequest, resp *interfaces.TaskStopResponse) error {
+	taskServices := h.getTaskServices()
```
Comment (notnoop) [inline client/allocrunner/taskrunner/service_hook.go]: Should this call `h.deregister()` or hold the lock like `Exited` and `PreKilling`?

File: client/allocrunner/taskrunner/service_hook.go [hunk 2]
```diff
@@ -16,6 +16,9 @@ import (
 	"github.com/hashicorp/nomad/plugins/drivers"
 )
 
+var _ interfaces.TaskExitedHook = &serviceHook{}
+var _ interfaces.TaskStopHook = &serviceHook{}
```
Comment (notnoop) [inline client/allocrunner/taskrunner/service_hook.go]: Worth adding other interfaces too, e.g. `TaskPreKillHook` anything else?

File: client/allocrunner/taskrunner/service_hook.go (hunk 1 above)
Comment (endocrimes) [reply client/allocrunner/taskrunner/service_hook.go]: fixed, yeah

File: client/allocrunner/taskrunner/service_hook.go (hunk 2 above)
Comment (endocrimes) [reply client/allocrunner/taskrunner/service_hook.go]: done

File: client/storage/interface.go [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package storage
```
Comment (notnoop) [inline client/storage/interface.go]: unrelated code?

File: client/allocrunner/taskrunner/task_runner_test.go [hunk 4]
```diff
@@ -2006,6 +2006,10 @@ func TestTaskRunner_UnregisterConsul_Retries(t *testing.T) {
 	// Removing canary and non-canary entries on retry
 	require.Equal(t, "remove", consulOps[4].Op)
 	require.Equal(t, "remove", consulOps[5].Op)
+
+	// Removing canary and non-canary entries on stop
+	require.Equal(t, "remove", consulOps[1].Op)
+	require.Equal(t, "remove", consulOps[2].Op)
```
Comment (notnoop) [inline client/allocrunner/taskrunner/task_runner_test.go]: Do you mean 6 and 7 here?

**Pull request merged**